### DISCUSSION TITLE
Typescript support

### DISCUSSION THREAD

**Discussion by LukeGT** [winkjs/wink-nlp#34] [New Features & Ideas]
Are there any plans to write Typescript declarations files for the wink modules? I use Typescript for all JS projects these days, and in order to use wink properly I've had to hack up some declaration files (I know there's the `any` type, but ew).

I'd be happy to contribute what I've written so far, but I can't promise it's complete or correct. It has been serving me well so far though :)

**Comment by sanjayaksaxena** [comment]
It will be great if you could contribute. We are not into Typescript, but we will fully support you in this endeavor. Let us discuss possibilities!

**Comment by sanjayaksaxena** [comment]
@LukeGT closing the issue for now, lets reopen whenever you are ready.

**Comment by pimpale** [comment]
Hello,

Thanks for this useful library. I have written a Typescript declaration file for wink-nlp and wink-eng-lite-web-mode.
You can view it here: https://gist.github.com/pimpale/9cf9ce5b0e903ddb3089fbf53f5c4bd0
These cover all of the public API that is documented on the website. (If I missed anything, please let me know.)

However, these types don't correctly represent the "defaulting" behavior that happens when you pass in an invalid itsHelper or asHelper, since I couldn't figure out how to represent that kind of behavior in Typescript.

I would like to contribute these files to the project, please let me know how to proceed.

Thanks, 
Govind Pimpale

**Reply by sanjayaksaxena** [reply]
Hello @pimpale 

Many thanks for your work. We would love to have your contributions. Please review our [contribution guidelines](https://github.com/winkjs/wink-nlp/blob/master/CONTRIBUTING.md).

We are not into Typescript but will surely review from the API coverage point of view and revert shortly. About default behaviors of `its.helper` in context of Typescript, we can do (re)search using google only!

How should we test these declarations? Please suggest.

— Sanjaya

**Reply by pimpale** [reply]
Hello,

I've reviewed the contribution guidelines, and I agree to the CLA. However, I could not find the form to sign that was mentioned in the guidelines.

With regards to testing, I think https://github.com/Microsoft/dtslint would be a good choice. I created a fork here, with dtslint testing enabled: https://github.com/pimpale/wink-nlp

To run it, enter `npm run dtslint`. 
The type declarations are stored in `types/index.d.ts`, and the tests are stored in `types/test.ts`.

Sincerely,
Govind Pimpale

**Reply by sanjayaksaxena** [reply]
Hello,

1. CLA can be signed at https://cla-assistant.io/winkjs/wink-nlp
2. API coverage looks complete.
3. Linting is perfect.
4. Wondering if `out<T>(itsf?: ItsFunction<T>): T;` & `out<T, U>(itsf?: ItsFunction<T>, asf?: AsFunction<T, U>): U;` would be a better representation? Remember we are not in to Typescript!
5. It will be nice if we execute the test file via chai/mocha combo — believe it is possible.
6. About default behaviour of `.out()`, we could not figure out anything.

Thanks once again for all your efforts.
— Sanjaya

**Reply by pimpale** [reply]
Hello,

1. Thanks, I signed it. You might want to include the link in the Contributor Guidelines document so that other people can find it easier.

4. Thats a good idea, I will add overloads.

5. I don't think it's possible to execute the test file that way. Chai and Mocha both actually run the code they're testing, whereas dtslint doesn't. The code in `tests.ts` is never actually evaluated. It's only typechecked for correctness. Therefore I don't think you could use Chai or Mocha for this purpose.

6. With regards to the default behavior, I actually think it would be a good idea to throw an error instead of use default behavior when the programmer passes in a disallowed method. 

The reason for this is that silently continuing with a different method than what the user requested will be major source of bugs. Rather than the programmer error being revealed when the `.out` function is called, the incorrectly formatted data could cause an error far away from the `.out` call. In addition, the programmer doesn't receive feedback that they used a disallowed method, since this behavior is silent. 

I think the correct behavior should be: Run default itsHelper and asHelper when the method provided is null. If it is an disallowed method, then throw an error saying that the method is not applicable for this collection.

This would be a breaking change, but the only code that would break is code that was not correctly using the API. I think it would be a good thing to consider for the next major version.

With regards to the current version, I think that the most bug free solution is:
* `out(): string[];`
* `out<T>(itsf: ItsFunction<T>): T[] | string[];`
* `out<T, U>(itsf: ItsFunction<T>, asf: AsFunction<T, U>): U | T[] | string[];`

This captures all possible failure modes. It will be a little verbose at times, but it is never wrong.

If the programmer is sure that they have the right method, they can write a wrapper function:
```tsx
// anything we can run an .out on
type OutApplicable =  Sentences | CustomEntities | SelectedCustomEntities | Entities | SelectedEntities | Tokens | SelectedTokens;

// only use if you're sure you have the right methods
function myOut<T, U>(toOut:OutApplicable, itsf: ItsFunction<T>, asf: AsFunction<T, U>): U {
  return (toOut.out(itsf, asf) as any) as  U
}

// usage example
const normalTokens = myOut(doc.tokens(), its.normal, as.array);
```

Please let me know of your thoughts.

Sincerely,
Govind Pimpale

**Reply by sanjayaksaxena** [reply]
Hi,

I agree with your suggestion of throwing error on invalid `its` or `as` helper. We have had a debate on this subject earlier as well. We will plan this a bit later. In the meantime, please send a PR for Typescript declarations.

Best,
— Sanjaya

**Reply by pimpale** [reply]
Thanks, I just sent my pull request.

Sincerely,
Govind Pimpale

**Reply by sanjayaksaxena** [reply]
Hi @pimpale 

I was trying to add support for `wink-eng-lite-model` in the types but I am not clear! Is it possible for you to review the [code](https://github.com/winkjs/wink-nlp/blob/9123624723ba67c15950e986c96ad22b80208e60/types/index.d.ts#L17) and guide. Idea is how to add support for multiple language models.

Thanks in advance.

Best,
Sanjaya

**Reply by pimpale** [reply]
Hi,

I took a look at the code for [wink-eng-lite-web-model](https://github.com/winkjs/wink-eng-lite-web-model), but I'm having trouble writing types for it last time since all the code is on one line with no indentation or formatting. Since it's in the dist folder, I assume it is the output of a build tool.

I looked at the wink-eng-lite-model (not web) repository, but I was unable to find any source code there either.

Would it be possible for you to upload the uncompiled source code so that I can write some types for it? 

Thanks,
Govind Pimpale

**Reply by sanjayaksaxena** [reply]
Hi,

I know... lotsa work to be done at our end! Contributor like you is such a great help. Until we release the source code after configuring the `.npmignore`, I am sharing a dropbox [link](https://www.dropbox.com/sh/gag8jx6fych50lr/AACMwNzvdorNM2bruWKfaicRa?dl=0) with you that contains the source code.

Thanks a lot for your support.
Sanjaya

**Reply by pimpale** [reply]
Hi,

Thank you for the dropbox link, it was very useful.

My solution to the problem was to create a Model type in wink-nlp that specifies the definition of a language model.
I then moved the wink-eng-lite-web-model types to the wink-eng-lite-web-model package itself. This way, if support for new languages is added, no changes will have to be made to the wink-nlp package,  only the model for the new language.

I have sent a pull request to this repository as well as wink-eng-lite-web-model with the necessary changes. I tested everything out with dtslint and it passes the tests successfully. 

I had to put a peer dependency for wink-nlp in wink-eng-lite-web-model, in order to import the Model type. So, the wink-eng-lite-web-model won't compile until a version of wink-nlp exporting the Model type is released. 

Let me know if you have any suggestions or comments.

Thanks,
Govind Pimpale

**Reply by sanjayaksaxena** [reply]
Thanks a ton @pimpale for all the effort and contribution. Would like to know that in web model under peer dependency you have mentioned wink-nlp v1.8.2:

```json
"peerDependencies": {
    "wink-nlp": "^1.8.2"
},
```

Is it because you are referring to the upcoming version on release of this PR? I noticed, when I was trying to run linting!

**Reply by pimpale** [reply]
Hi,

That is what I was trying to do, since the current version (1.8.1) doesn't yet export Model and ModelAddon. For linting purposes you can edit it to depend on the current state git repository like so:

```json
"peerDependencies": {
    "wink-nlp": "winkjs/wink-nlp"
},
```
And then try linting like that. 

Sincerely,
Govind Pimpale

**Reply by sanjayaksaxena** [reply]
Yes, I did that only — thanks once again. Will release in a couple of days and also plan to release the source along with the distribution of model after adding source directory in `.npmignore`.

**Reply by pimpale** [reply]
Ok sounds good. Thanks for accepting the PRs!

Thanks,
Govind Pimpale

**Reply by sanjayaksaxena** [reply]
Thanks again @pimpale, have released the new versions of wink-nlp & wink-eng-lite-web-model. 🥳

**Comment by sanjayaksaxena** [comment]
Many thanks @pimpale for your contribution — have released the new version 1.8.0. 🎉

**Reply by pimpale** [reply]
Thanks! Glad I was able to help! :)

**Comment by sanjayaksaxena** [comment]
Hi @pimpale 

Have released a new version v1.10.0 that adds model loading feature to bm25 vectorizer. I have updated the typescript for the same and request you to review it please.

Thanks a lot.

— Sanjaya

**Reply by pimpale** [reply]
Hi, 

It looks good to me.

Thanks,
Govind Pimpale

**Comment by dcsan** [comment]
how are the typescript defs to be imported?

```
npm i --save-dev @types/wink-nlp
npm ERR! code E404
npm ERR! 404 Not Found - GET https://registry.npmjs.org/@types%2fwink-nlp - Not found
```

**Comment by rachnachakraborty** [comment]
Hi @dcsan 

Please check out the resolution in [ issue#25 ](https://github.com/winkjs/wink-nlp/issues/125) from @itPiligrim 

Hope this works for you.

Thanks,
Rachna

**Comment by winkmod-apk** [comment]
It would be great to have your contribution. While TypeScript isn’t our main focus right now, we’re ready to support you in moving forward. Communities like [winkmodapk](https://www.winkmodapk.com.co) also grow through collaboration and shared ideas — let’s discuss the possibilities together.
